compositor_pipeline/src/queue.rs

```diff
@@ -19,58 +21,100 @@ pub enum QueueError {
+    clock_start: Instant,
 }
 
 impl Queue {
     pub fn new(output_framerate: Framerate) -> Self {
         Queue {
-            internal_queue: Mutex::new(InternalQueue::new(output_framerate)),
+            internal_queue: Mutex::new(InternalQueue::new()),
             check_queue_channel: unbounded(),
             output_framerate,
             buffer_duration: DEFAULT_BUFFER_DURATION,
+            clock_start: Instant::now(),
         }
     }
 
     pub fn add_input(&self, input_id: InputId) {
-        let mut internal_queue = self.internal_queue.lock().unwrap();
-        internal_queue.add_input(input_id);
+        self.internal_queue.lock().unwrap().add_input(input_id);
     }
 
-    #[allow(dead_code)]
     pub fn remove_input(&self, input_id: &InputId) {
-        let mut internal_queue = self.internal_queue.lock().unwrap();
-        // TODO: gracefully remove input - wait until last enqueued frame PTS is smaller than output PTS
-        internal_queue.remove_input(input_id);
+        self.internal_queue.lock().unwrap().remove_input(input_id);
     }
 
     pub fn start(self: &Arc<Self>, sender: Sender<FrameSet<InputId>>) {
         let queue = self.clone();
         let tick_duration = self.output_framerate.get_interval_duration();
-        let buffer_duration = self.buffer_duration;
 
         QueueThread::new(
             queue,
             sender,
             queue_thread::Options {
-                buffer_duration,
                 tick_duration,
+                buffer_duration: self.buffer_duration,
+                output_framerate: self.output_framerate,
+                clock_start: self.clock_start,
             },
         )
         .spawn();
     }
 
     pub fn enqueue_frame(&self, input_id: InputId, frame: Frame) -> Result<(), QueueError> {
+        let is_first_frame = self
+            .internal_queue
+            .lock()
+            .unwrap()
+            .input_offset(&input_id)
+            .is_none();
+        if is_first_frame {
+            // Sleep here ensures that we will buffer `self.buffer_duration` on each input.
+            // It also makes calculation easier because PTS of frames will be already offset
+            // by a correct value.
+            thread::sleep(self.buffer_duration);
+        }
+
         let mut internal_queue = self.internal_queue.lock().unwrap();
 
-        internal_queue.enqueue_frame(input_id.clone(), frame)?;
-        internal_queue.drop_input_useless_frames(input_id)?;
+        internal_queue.enqueue_frame(input_id.clone(), frame, self.clock_start)?;
+
+        // We don't know when pipeline is started, so we can't resolve real_next_pts,
+        // but we can remove frames based on estimated PTS. This only works if queue
+        // is able to push frames in real time and is never behind more than one frame.
+        let framerate_tick = Duration::from_secs_f64(1.0 / self.output_framerate.0 as f64);
+        let estimated_pts = self.clock_start.elapsed() - framerate_tick;
```

Comment:
I'm not really sure if one tick is enough, technically queue does not do anything heavy, so I don't see a reason why it would be behind with pushing frames, but if we want to be safe we could add e.g.

```suggestion
        let estimated_pts = self.clock_start.elapsed() - framerate_tick * 10;
```

This drop_frame is not strictly necessary, when pipeline is running because queue thread will clean it up, but if pipeline did not start yet that cleanup in QueueThread would not be scheduled